File: tests/csharp/oop_advanced/tostring_override.cs
// vybe-test: csharp/oop_advanced/tostring_override
// origin: languages/csharp/tests/csharp/test_oop_advanced.rs

string __buf = "";

void __P(string s) {
    __buf = __buf + s + "\n";
}

void __Pr(string s) {
    __buf = __buf + s;
}

// The final WriteLine contributes a trailing newline that the expected line
// vector never carried, so BOTH forms are accepted.
void __Check(string want) {
    if (__buf != want && __buf != want + "\n") {
        Console.WriteLine("FAIL: want [" + want + "] got [" + __buf + "]");
        throw new Exception("assertion failed");
    }
}

class Person {
    public string Name { get; set; }
    public int Age { get; set; }
    public Person(string name, int age) { Name = name; Age = age; }
    public override string ToString() { return Name + " (" + Age + ")"; }
}
var p = new Person("Alice", 30);
__P((p.ToString()).ToString());
__P((p).ToString());
__Check("Alice (30)\nAlice (30)");

File: tests/js/class_patterns/tostring_override.js
// vybe-test: js/class_patterns/tostring_override
// origin: languages/js/tests/js/test_class_patterns.rs

function __line(...args) {
    // console.log joins its arguments with a single space. String() is the
    // coercion Vybe's logging host applies to each one.
    return args.map(String).join(" ");
}

function __check(got, want) {
    if (got !== want) {
        console.log("FAIL: want [" + want + "] got [" + got + "]");
        throw new Error("assertion failed");
    }
}

class Money {
    constructor(amount, currency) {
        this.amount = amount;
        this.currency = currency;
    }
    toString() { return this.amount + " " + this.currency; }
}
let m = new Money(100, "USD");
__check(__line("" + m), "100 USD");
__check(__line(`${m}`), "100 USD");
